src/utilities/ModerationUtility.ts

```diff
@@ -0,0 +1,205 @@
+
+export type Case = Omit<Moderation, "createdAt" | "caseId" | "modLogMessageId">;
+export type CaseWithReference = Omit<Moderation, "action"> & { action: CaseAction | "Punishment Expiry"; caseReference: Moderation | null };
+
+export const LogChannelNames = ["modlogs", "modlog", "mod-log", "mod-logs", "logs", "sentry-logs", "sentry-log"];
+
+export class ModerationUtility extends Utility {
+  public constructor(context: Utility.Context, options: Utility.Options) {
+    super(context, {
+      ...options,
+      name: "moderation",
+    });
+  }
+
+  public async createCase(guild: Guild, data: Case, dm = true): Promise<Result<[CaseWithReference, APIEmbed], Error>> {
+    await this.container.prisma.guild.upsert({ create: { id: guild.id }, update: {}, where: { id: guild.id } });
+
+    const modCase = await this.container.prisma.moderation.create({
+      data: {
+        ...data,
+        caseId: await this.generateCaseId(guild.id),
+      },
+      include: {
+        caseReference: true,
+      },
+    });
+
+    if (data.duration) {
+      await this.container.tasks.create("expiringCase", { id: modCase.caseId }, data.duration);
+    }
+
+    const moderator = await this.container.client.users.fetch(data.moderatorId);
+    const embedResult = await this.sendModLogMessage(guild, moderator, modCase);
+
+    const embed = embedResult.isOk() ? embedResult.unwrap() : embedResult.unwrapErr()[1];
+
+    if (dm && data.action !== "Unban") {
+      try {
+        const user = await this.container.client.users.fetch(data.userId);
+
+        const row = new ActionRowBuilder<ButtonBuilder>().addComponents(
+          new ButtonBuilder().setCustomId("void").setDisabled(true).setStyle(ButtonStyle.Primary).setLabel(`Sent from ${guild.name}`)
+        );
+
+        await user.send({ embeds: [embed], components: [row] });
+      } catch (_err) {}
+    }
+
+    try {
+      switch (data.action) {
+        case "Warn":
+          break;
+        case "Timeout":
+          await guild.members.edit(data.userId, { communicationDisabledUntil: new Date(Date.now() + data.duration!), reason: data.reason });
+          break;
+        case "Kick":
+          await guild.members.kick(data.userId, data.reason);
+          break;
+        case "Softban":
+          await guild.members.ban(data.userId, { deleteMessageSeconds: (Time.Day * 7) / Time.Second, reason: data.reason });
+          await guild.bans.remove(data.userId, "Removing ban as part of softban");
+          break;
+        case "Ban":
+          await guild.members.ban(data.userId, { deleteMessageSeconds: (Time.Day * 7) / Time.Second, reason: data.reason });
+          break;
+        case "Unban":
+          await guild.bans.remove(data.userId, "Removing ban as part of softban");
+          break;
+      }
+    } catch (error) {
+      this.container.logger.error(error);
+
+      captureException(error);
+
+      return Result.err(error as Error);
+    }
+
+    return Result.ok([modCase, embed]);
+  }
```

Comment:
Reminder to myself to not push this until i implement permissions checks

Currently no heirarchy is established, and I can punish myself, other bots, etc

Disable this before pushing

Reply:
resolved in be51d4bfd8b0eb07d259f6eea8f0f9f97324b07b